bug node V-101 is marked clean but hash is dirty discovered by fuzzing

Following bug can be triggered on this branch: `kjezek/bug-clean-node-dirty-hash`: 

Running of
```
  go test  -run=FuzzArchiveTrie_RandomAccountStorageOps/6992b1b118b505e6 

```
triggers 
```
--- FAIL: FuzzArchiveTrie_RandomAccountStorageOps (0.71s)
    --- FAIL: FuzzArchiveTrie_RandomAccountStorageOps/6992b1b118b505e6 (0.70s)
        archive_trie_fuzzing_test.go:972: trie verification fails: 
            node V-101 is marked clean but hash is dirty

```

--- a/go/state/mpt/forest_test.go
+++ b/go/state/mpt/forest_test.go
@@ -827,3 +827,65 @@ func TestForest_NodeHandlingDoesNotDeadlock(t *testing.T) {
 		t.Errorf("failed to close forest: %v", err)
 	}
 }
+
+func TestForest_CheckPassesAfterReopeningDirectory(t *testing.T) {
+	for _, config := range allMptConfigs {
+		t.Run(config.Name, func(t *testing.T) {
+			for _, mode := range []StorageMode{Immutable, Mutable} {
+				t.Run(mode.String(), func(t *testing.T) {
+					// Create a forest, fill it with data, check it, close it, reopen
+					// it, and check that the consistency test still passes.
+					dir := t.TempDir()
+					forestConfig := ForestConfig{
+						Mode:          mode,
+						CacheCapacity: 1024,
+					}
+					forest, err := OpenFileForest(dir, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to open forest: %v", err)
+					}
+
+					// Fill the forest with some data.
+					const N = 10
+					root := NewNodeReference(EmptyId())
+					for a := 0; a < N; a++ {
+						addr := common.Address{byte(a)}
+						root, err = forest.SetAccountInfo(&root, addr, AccountInfo{Nonce: common.Nonce{1}})
+						if err != nil {
+							t.Fatalf("failed to create account %v: %v", addr, err)
+						}
+						for k := 0; k < N; k++ {
+							root, err = forest.SetValue(&root, addr, common.Key{byte(k)}, common.Value{byte(k)})
+							if err != nil {
+								t.Fatalf("failed to update value %v: %v", addr, err)
+							}
+						}
+					}
+					_, _, err = forest.updateHashesFor(&root)
+					if err != nil {
+						t.Fatalf("failed to update hashes: %v", err)
+					}
+					if err := forest.Check(&root); err != nil {
+						t.Fatalf("check for initial forest failed: %v", err)
+					}
+
+					if err := forest.Close(); err != nil {
+						t.Fatalf("failed to close forest: %v", err)
+					}
+
+					// Reopen the forest and run check again.
+					forest, err = OpenFileForest(dir, config, forestConfig)
+					if err != nil {
+						t.Fatalf("failed to re-open forest: %v", err)
+					}
+					if err := forest.Check(&root); err != nil {
+						t.Fatalf("check for re-opened forest failed: %v", err)
+					}
+					if err := forest.Close(); err != nil {
+						t.Fatalf("failed to close forest: %v", err)
+					}
+				})
+			}
+		})
+	}
+}

--- a/go/state/mpt/hasher_test.go
+++ b/go/state/mpt/hasher_test.go
@@ -43,7 +43,7 @@ func TestHasher_ExtensionNode_GetHash_DirtyHashesAreIgnored(t *testing.T) {
 			// The dirty flag is cleared.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if !handle.Get().(*ExtensionNode).hashDirty {
+			if handle.Get().(*ExtensionNode).hasCleanHash() {
 				t.Errorf("dirty hash flag should not be cleared")
 			}
 			if !handle.Get().(*ExtensionNode).nextHashDirty {
@@ -80,7 +80,7 @@ func TestHasher_ExtensionNode_UpdateHash_DirtyHashesAreRefreshed(t *testing.T) {
 			// The dirty flag is cleared.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if handle.Get().(*ExtensionNode).hashDirty {
+			if !handle.Get().(*ExtensionNode).hasCleanHash() {
 				t.Errorf("dirty hash flag should be cleared")
 			}
 			if handle.Get().(*ExtensionNode).nextHashDirty {
@@ -114,7 +114,7 @@ func TestHasher_BranchNode_GetHash_DirtyHashesAreIgnored(t *testing.T) {
 			// The dirty flag is not touched.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if !handle.Get().(*BranchNode).hashDirty {
+			if handle.Get().(*BranchNode).hasCleanHash() {
 				t.Errorf("dirty hash flag should not be cleared")
 			}
 			if handle.Get().(*BranchNode).dirtyHashes != ((1 << 0x7) | (1 << 0xd)) {
@@ -150,7 +150,7 @@ func TestHasher_BranchNode_UpdateHash_DirtyHashesAreRefreshed(t *testing.T) {
 			// The dirty flag is cleared.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if handle.Get().(*BranchNode).hashDirty {
+			if !handle.Get().(*BranchNode).hasCleanHash() {
 				t.Errorf("dirty hash flag should be cleared")
 			}
 			if handle.Get().(*BranchNode).dirtyHashes != 0 {
@@ -184,7 +184,7 @@ func TestHasher_BranchNode_UpdateHash_DirtyFlagsForEmptyChildrenAreClearedButNoU
 			// The dirty flag is cleared.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if handle.Get().(*BranchNode).hashDirty {
+			if !handle.Get().(*BranchNode).hasCleanHash() {
 				t.Errorf("dirty hash flag should be cleared")
 			}
 			if handle.Get().(*BranchNode).dirtyHashes != 0 {
@@ -216,7 +216,7 @@ func TestHasher_AccountNode_GetHash_DirtyHashesAreIgnored(t *testing.T) {
 			// The dirty flag is not changed.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if !handle.Get().(*AccountNode).hashDirty {
+			if handle.Get().(*AccountNode).hasCleanHash() {
 				t.Errorf("dirty hash flags should not be changed")
 			}
 			if !handle.Get().(*AccountNode).storageHashDirty {
@@ -248,7 +248,7 @@ func TestHasher_AccountNode_UpdateHash_DirtyHashesAreRefreshed(t *testing.T) {
 			// The dirty flag is cleared.
 			handle := node.GetReadHandle()
 			defer handle.Release()
-			if handle.Get().(*AccountNode).hashDirty {
+			if !handle.Get().(*AccountNode).hasCleanHash() {
 				t.Errorf("dirty hash flag should be cleared")
 			}
 			if handle.Get().(*AccountNode).storageHashDirty {

--- a/go/state/mpt/nodes.go
+++ b/go/state/mpt/nodes.go
@@ -368,19 +368,46 @@ func (c *nodeCheckContext) isCompatible(other *nodeCheckContext) bool {
 
 // nodeBase is an optional common base type for nodes.
 type nodeBase struct {
-	hash      common.Hash // the hash of this node (may be dirty)
-	hashDirty bool        // indicating whether this node's hash is dirty
-	clean     bool        // by default nodes are dirty (clean == false)
-	frozen    bool        // a flag marking the node as immutable (default: mutable)
+	hash       common.Hash // the hash of this node (may be dirty)
+	hashStatus hashStatus  // indicating whether this node's hash is valid
+	clean      bool        // by default nodes are dirty (clean == false)
+	frozen     bool        // a flag marking the node as immutable (default: mutable)
+}
+
+type hashStatus byte
+
+const (
+	hashStatusClean   hashStatus = 0 // the hash is up-to-date, matching the nodes content
+	hashStatusDirty   hashStatus = 1 // the hash is out-of-date and needs to be refreshed
+	hashStatusUnknown hashStatus = 2 // the hash is missing / invalid
+)
+
+func (s hashStatus) String() string {
+	switch s {
+	case hashStatusClean:
+		return "clean"
+	case hashStatusDirty:
+		return "dirty"
+	default:
+		return "unknown"
+	}
 }
 
 func (n *nodeBase) GetHash() (common.Hash, bool) {
-	return n.hash, n.hashDirty
+	return n.hash, n.hashStatus != hashStatusClean
 }
 
 func (n *nodeBase) SetHash(hash common.Hash) {
 	n.hash = hash
-	n.hashDirty = false
+	n.hashStatus = hashStatusClean
+}
+
+func (n *nodeBase) hasCleanHash() bool {
+	return n.hashStatus == hashStatusClean
+}
+
+func (n *nodeBase) getHashStatus() hashStatus {
+	return n.hashStatus
 }
 
 func (n *nodeBase) IsFrozen() bool {
@@ -405,12 +432,25 @@ func (n *nodeBase) MarkClean() {
 
 func (n *nodeBase) markDirty() {
 	n.clean = false
+	n.hashStatus = hashStatusDirty
 }
 
 func (n *nodeBase) Release() {
 	// The node is disconnected from the disk version and thus clean.
 	// TODO: test that all released nodes are clean!
 	n.clean = true
+	n.hashStatus = hashStatusClean
+}
+
+func (n *nodeBase) check(thisRef *NodeReference) error {
+	var errs []error
+	if !n.IsDirty() && n.hashStatus == hashStatusDirty {
+		errs = append(errs, fmt.Errorf("node %v is marked clean but hash is dirty", thisRef.Id()))
+	}
+	if n.IsDirty() && n.hashStatus == hashStatusUnknown {
+		errs = append(errs, fmt.Errorf("node %v is marked dirty but hash is marked unknown (should be dirty or clean)", thisRef.Id()))
+	}
+	return errors.Join(errs...)
 }
 
 // ----------------------------------------------------------------------------
@@ -444,7 +484,7 @@ func (e EmptyNode) SetAccount(manager NodeManager, thisRef *NodeReference, this 
 	}
 	defer handle.Release()
 	res := handle.Get().(*AccountNode)
-	res.hashDirty = true
+	res.markDirty()
 	res.address = address
 	res.info = info
 	res.pathLength = byte(len(path))
@@ -463,7 +503,7 @@ func (e EmptyNode) SetValue(manager NodeManager, thisRef *NodeReference, this sh
 	res := handle.Get().(*ValueNode)
 	res.key = key
 	res.value = value
-	res.hashDirty = true
+	res.markDirty()
 	res.pathLength = byte(len(path))
 	return ref, true, nil
 }
@@ -602,7 +642,6 @@ func (n *BranchNode) setNextNode(
 	if newRoot.Id() == child.Id() {
 		if hasChanged {
 			n.markDirty()
-			n.hashDirty = true
 			n.markChildHashDirty(byte(path[0]))
 		}
 		return *thisRef, hasChanged, nil
@@ -627,7 +666,6 @@ func (n *BranchNode) setNextNode(
 
 	wasEmpty := child.Id().IsEmpty()
 	n.children[path[0]] = newRoot
-	n.hashDirty = true
 	n.markChildHashDirty(byte(path[0]))
 	n.setChildFrozen(byte(path[0]), false)
 
@@ -675,7 +713,6 @@ func (n *BranchNode) setNextNode(
 				}
 
 				extensionNode.path.Prepend(remainingPos)
-				extensionNode.hashDirty = true
 				extensionNode.markDirty()
 			} else if remaining.Id().IsBranch() {
 				// An extension needs to replace this branch.
@@ -687,7 +724,6 @@ func (n *BranchNode) setNextNode(
 				extension := handle.Get().(*ExtensionNode)
 				extension.path = SingleStepPath(remainingPos)
 				extension.next = remaining
-				extension.hashDirty = true
 				extension.nextHashDirty = n.isChildHashDirty(byte(remainingPos))
 				if !extension.nextHashDirty {
 					// TODO: add unit test coverage for this!
@@ -816,13 +852,14 @@ func (n *BranchNode) Check(source NodeSource, thisRef *NodeReference, _ []Nibble
 	//  - non-dirty hashes for child nodes are valid
 	//  - mask of frozen children is consistent
 	numChildren := 0
-	errs := []error{}
+	var errs []error
 
-	// TODO: move to base node check
-	if !n.IsDirty() && n.hashDirty {
-		errs = append(errs, fmt.Errorf("node %v is marked clean but hash is dirty", thisRef.Id()))
+	if err := n.nodeBase.check(thisRef); err != nil {
+		errs = append(errs, err)
 	}
-	if !n.hashDirty && n.dirtyHashes != 0 {
+
+	hashWithParent := source.getConfig().HashStorageLocation == HashStoredWithParent
+	if hashWithParent && n.hasCleanHash() && n.dirtyHashes != 0 {
 		errs = append(errs, fmt.Errorf("node %v is has clean hash but child hashes are dirty: %016b", thisRef.Id(), n.dirtyHashes))
 	}
 
@@ -865,7 +902,7 @@ func (n *BranchNode) Check(source NodeSource, thisRef *NodeReference, _ []Nibble
 
 func (n *BranchNode) Dump(out io.Writer, source NodeSource, thisRef *NodeReference, indent string) error {
 	errs := []error{}
-	fmt.Fprintf(out, "%sBranch (ID: %v, dirty: %t, frozen: %t, Dirty: %016b, Embedded: %016b, Frozen: %016b, Hash: %v, dirtyHash: %t):\n", indent, thisRef.Id(), n.IsDirty(), n.IsFrozen(), n.dirtyHashes, n.embeddedChildren, n.frozenChildren, formatHashForDump(n.hash), n.hashDirty)
+	fmt.Fprintf(out, "%sBranch (ID: %v, dirty: %t, frozen: %t, Dirty: %016b, Embedded: %016b, Frozen: %016b, Hash: %v, hashState: %v):\n", indent, thisRef.Id(), n.IsDirty(), n.IsFrozen(), n.dirtyHashes, n.embeddedChildren, n.frozenChildren, formatHashForDump(n.hash), n.getHashStatus())
 	for i, child := range n.children {
 		if child.Id().IsEmpty() {
 			continue
@@ -1009,7 +1046,6 @@ func (n *ExtensionNode) GetSlot(source NodeSource, address common.Address, path 
 func (n *ExtensionNode) setNextNode(
 	manager NodeManager,
 	thisRef *NodeReference,
-	this shared.WriteHandle[Node],
 	path []Nibble,
 	valueIsEmpty bool,
 	createSubTree func(*NodeReference, shared.WriteHandle[Node], []Nibble) (NodeReference, bool, error),
@@ -1044,7 +1080,7 @@ func (n *ExtensionNode) setNextNode(
 				*newNode = *n
 				newNode.markDirty()
 				newNode.markMutable()
-				thisRef, this, n = &newRef, handle, newNode
+				thisRef, n = &newRef, newNode
 				isClone = true
 			}
 
@@ -1061,7 +1097,6 @@ func (n *ExtensionNode) setNextNode(
 				extension := handle.Get().(*ExtensionNode)
 				n.path.AppendAll(&extension.path)
 				n.next = extension.next
-				n.hashDirty = true
 				n.nextHashDirty = extension.nextHashDirty
 				if !extension.nextHashDirty {
 					// TODO: add unit test coverage for this!
@@ -1075,7 +1110,6 @@ func (n *ExtensionNode) setNextNode(
 				}
 			} else if newRoot.Id().IsBranch() {
 				n.next = newRoot
-				n.hashDirty = true
 				n.nextHashDirty = true
 				n.markDirty()
 			} else {
@@ -1108,7 +1142,6 @@ func (n *ExtensionNode) setNextNode(
 			}
 		} else if hasChanged {
 			n.markDirty()
-			n.hashDirty = true
 			n.nextHashDirty = true
 		}
 		return *thisRef, hasChanged, err
@@ -1131,7 +1164,7 @@ func (n *ExtensionNode) setNextNode(
 		*newNode = *n
 		newNode.markDirty()
 		newNode.markMutable()
-		thisRef, this, n = &newRef, handle, newNode
+		thisRef, n = &newRef, newNode
 		isClone = true
 	}
 
@@ -1159,7 +1192,6 @@ func (n *ExtensionNode) setNextNode(
 		branch.children[n.path.Get(commonPrefixLength)] = *thisRef
 		branch.markChildHashDirty(byte(n.path.Get(commonPrefixLength)))
 		n.path.ShiftLeft(commonPrefixLength + 1)
-		n.hashDirty = true
 		n.markDirty()
 		thisNodeWasReused = true
 	} else {
@@ -1179,8 +1211,8 @@ func (n *ExtensionNode) setNextNode(
 		// Reuse current node unless already taken.
 		extension := n
 		extensionRef := *thisRef
-		extensionHandle := this
 		if thisNodeWasReused {
+			var extensionHandle shared.WriteHandle[Node]
 			extensionRef, extensionHandle, err = manager.createExtension()
 			if err != nil {
 				return NodeReference{}, false, err
@@ -1193,7 +1225,6 @@ func (n *ExtensionNode) setNextNode(
 
 		extension.path = CreatePathFromNibbles(path[0:commonPrefixLength])
 		extension.next = branchRef
-		extension.hashDirty = true
 		extension.nextHashDirty = true
 		extension.markDirty()
 		newRoot = extensionRef
@@ -1215,7 +1246,7 @@ func (n *ExtensionNode) setNextNode(
 }
 
 func (n *ExtensionNode) SetAccount(manager NodeManager, thisRef *NodeReference, this shared.WriteHandle[Node], address common.Address, path []Nibble, info AccountInfo) (NodeReference, bool, error) {
-	return n.setNextNode(manager, thisRef, this, path, info.IsEmpty(),
+	return n.setNextNode(manager, thisRef, path, info.IsEmpty(),
 		func(next *NodeReference, node shared.WriteHandle[Node], path []Nibble) (NodeReference, bool, error) {
 			return node.Get().SetAccount(manager, next, node, address, path, info)
 		},
@@ -1223,7 +1254,7 @@ func (n *ExtensionNode) SetAccount(manager NodeManager, thisRef *NodeReference, 
 }
 
 func (n *ExtensionNode) SetValue(manager NodeManager, thisRef *NodeReference, this shared.WriteHandle[Node], key common.Key, path []Nibble, value common.Value) (NodeReference, bool, error) {
-	return n.setNextNode(manager, thisRef, this, path, value == (common.Value{}),
+	return n.setNextNode(manager, thisRef, path, value == (common.Value{}),
 		func(next *NodeReference, node shared.WriteHandle[Node], path []Nibble) (NodeReference, bool, error) {
 			return node.Get().SetValue(manager, next, node, key, path, value)
 		},
@@ -1231,7 +1262,7 @@ func (n *ExtensionNode) SetValue(manager NodeManager, thisRef *NodeReference, th
 }
 
 func (n *ExtensionNode) SetSlot(manager NodeManager, thisRef *NodeReference, this shared.WriteHandle[Node], address common.Address, path []Nibble, key common.Key, value common.Value) (NodeReference, bool, error) {
-	return n.setNextNode(manager, thisRef, this, path, true,
+	return n.setNextNode(manager, thisRef, path, true,
 		func(next *NodeReference, node shared.WriteHandle[Node], path []Nibble) (NodeReference, bool, error) {
 			return node.Get().SetSlot(manager, next, node, address, path, key, value)
 		},
@@ -1239,7 +1270,7 @@ func (n *ExtensionNode) SetSlot(manager NodeManager, thisRef *NodeReference, thi
 }
 
 func (n *ExtensionNode) ClearStorage(manager NodeManager, thisRef *NodeReference, this shared.WriteHandle[Node], address common.Address, path []Nibble) (newRoot NodeReference, hasChanged bool, err error) {
-	return n.setNextNode(manager, thisRef, this, path, true,
+	return n.setNextNode(manager, thisRef, path, true,
 		func(next *NodeReference, node shared.WriteHandle[Node], path []Nibble) (NodeReference, bool, error) {
 			return node.Get().ClearStorage(manager, next, node, address, path)
 		},
@@ -1282,12 +1313,14 @@ func (n *ExtensionNode) Check(source NodeSource, thisRef *NodeReference, _ []Nib
 	//  - extension can only be followed by a branch
 	//  - hash of sub-tree is either dirty or correct
 	//  - frozen flags are consistent
-	errs := []error{}
+	var errs []error
 
-	if !n.IsDirty() && n.hashDirty {
-		errs = append(errs, fmt.Errorf("node %v is marked clean but hash is dirty", thisRef.Id()))
+	if err := n.nodeBase.check(thisRef); err != nil {
+		errs = append(errs, err)
 	}
-	if !n.hashDirty && n.nextHashDirty {
+
+	hashWithParent := source.getConfig().HashStorageLocation == HashStoredWithParent
+	if hashWithParent && n.hasCleanHash() && n.nextHashDirty {
 		errs = append(errs, fmt.Errorf("node %v is marked to have a clean hash but next hash is dirty", thisRef.Id()))
 	}
 
@@ -1324,7 +1357,7 @@ func (n *ExtensionNode) Check(source NodeSource, thisRef *NodeReference, _ []Nib
 
 func (n *ExtensionNode) Dump(out io.Writer, source NodeSource, thisRef *NodeReference, indent string) error {
 	errs := []error{}
-	fmt.Fprintf(out, "%sExtension (ID: %v/%t, dirtyHash: %t, Embedded: %t, Hash: %v, dirtyHash: %t): %v\n", indent, thisRef.Id(), n.IsFrozen(), n.nextHashDirty, n.nextIsEmbedded, formatHashForDump(n.hash), n.hashDirty, &n.path)
+	fmt.Fprintf(out, "%sExtension (ID: %v/%t, nextHashDirty: %t, Embedded: %t, Hash: %v, hashState: %v): %v\n", indent, thisRef.Id(), n.IsFrozen(), n.nextHashDirty, n.nextIsEmbedded, formatHashForDump(n.hash), n.getHashStatus(), &n.path)
 	if handle, err := source.getViewAccess(&n.next); err == nil {
 		defer handle.Release()
 		if err := handle.Get().Dump(out, source, &n.next, indent+"  "); err != nil {
@@ -1441,12 +1474,10 @@ func (n *AccountNode) SetAccount(manager NodeManager, thisRef *NodeReference, th
 			newNode.markDirty()
 			newNode.markMutable()
 			newNode.info = info
-			newNode.hashDirty = true
 			return newRef, false, nil
 		}
 
 		n.info = info
-		n.hashDirty = true
 		n.markDirty()
 		return *thisRef, true, nil
 	}
@@ -1465,7 +1496,7 @@ func (n *AccountNode) SetAccount(manager NodeManager, thisRef *NodeReference, th
 	sibling := handle.Get().(*AccountNode)
 	sibling.address = address
 	sibling.info = info
-	sibling.hashDirty = true
+	sibling.markDirty()
 
 	thisPath := AddressToNibblePath(n.address, manager)
 	newRoot, err := splitLeafNode(manager, thisRef, thisPath[:], n, this, path, &siblingRef, sibling, handle)
@@ -1515,7 +1546,6 @@ func splitLeafNode(
 
 		extension.path = CreatePathFromNibbles(siblingPath[0:commonPrefixLength])
 		extension.next = branchRef
-		extension.hashDirty = true
 		extension.nextHashDirty = true
 		extension.markDirty()
 	}
@@ -1539,7 +1569,7 @@ func splitLeafNode(
 	branch.children[partialPath[commonPrefixLength]] = *thisRef
 	branch.children[siblingPath[commonPrefixLength]] = *siblingRef
 	branch.markChildHashDirty(byte(siblingPath[commonPrefixLength]))
-	branch.hashDirty = true
+	branch.markDirty()
 
 	// Update hash if present.
 	if hash, dirty := this.GetHash(); thisModified || dirty {
@@ -1599,16 +1629,13 @@ func (n *AccountNode) SetSlot(manager NodeManager, thisRef *NodeReference, this 
 			newNode.markMutable()
 			newNode.storage = root
 			newNode.storageHashDirty = true
-			newNode.hashDirty = true
 			return newRef, false, nil
 		}
 		n.storage = root
 		n.storageHashDirty = true
-		n.hashDirty = true
 		n.markDirty()
 		hasChanged = true
 	} else if hasChanged {
-		n.hashDirty = true
 		n.storageHashDirty = true
 		n.markDirty()
 	}
@@ -1634,7 +1661,6 @@ func (n *AccountNode) ClearStorage(manager NodeManager, thisRef *NodeReference, 
 		newNode.markMutable()
 		newNode.storage = NewNodeReference(EmptyId())
 		newNode.storageHashDirty = true
-		newNode.hashDirty = true
 		return newRef, false, nil
 	}
 
@@ -1644,7 +1670,6 @@ func (n *AccountNode) ClearStorage(manager NodeManager, thisRef *NodeReference, 
 
 	n.storage = NewNodeReference(EmptyId())
 	n.markDirty()
-	n.hashDirty = true
 	n.storageHashDirty = true
 	return *thisRef, true, err
 }
@@ -1683,11 +1708,9 @@ func (n *AccountNode) setPathLength(manager NodeManager, thisRef *NodeReference,
 		newNode.markDirty()
 		newNode.markMutable()
 		newNode.pathLength = length
-		newNode.hashDirty = true
 		return newRef, false, nil
 	}
 
-	n.hashDirty = true
 	n.pathLength = length
 	n.markDirty()
 	return *thisRef, true, nil
@@ -1712,12 +1735,14 @@ func (n *AccountNode) Check(source NodeSource, thisRef *NodeReference, path []Ni
 	//  - the account is at a correct position in the trie
 	//  - frozen flags are consistent
 	//  - path length
-	errs := []error{}
+	var errs []error
 
-	if !n.IsDirty() && n.hashDirty {
-		errs = append(errs, fmt.Errorf("node %v is marked clean but hash is dirty", thisRef.Id()))
+	if err := n.nodeBase.check(thisRef); err != nil {
+		errs = append(errs, err)
 	}
-	if !n.hashDirty && n.storageHashDirty {
+
+	hashWithParent := source.getConfig().HashStorageLocation == HashStoredWithParent
+	if hashWithParent && n.hasCleanHash() && n.storageHashDirty {
 		errs = append(errs, fmt.Errorf("node %v is marked to have a clean hash but storage hash is dirty", thisRef.Id()))
 	}
 
@@ -1758,7 +1783,7 @@ func (n *AccountNode) Check(source NodeSource, thisRef *NodeReference, path []Ni
 
 func (n *AccountNode) Dump(out io.Writer, source NodeSource, thisRef *NodeReference, indent string) error {
 	errs := []error{}
-	fmt.Fprintf(out, "%sAccount (ID: %v, dirty: %t, frozen: %t, path length: %v, Hash: %v, dirtyHash: %t): %v - %v\n", indent, thisRef.Id(), n.IsDirty(), n.IsFrozen(), n.pathLength, formatHashForDump(n.hash), n.hashDirty, n.address, n.info)
+	fmt.Fprintf(out, "%sAccount (ID: %v, dirty: %t, frozen: %t, path length: %v, Hash: %v, hashState: %v): %v - %v\n", indent, thisRef.Id(), n.IsDirty(), n.IsFrozen(), n.pathLength, formatHashForDump(n.hash), n.getHashStatus(), n.address, n.info)
 	if n.storage.Id().IsEmpty() {
 		return nil
 	}
@@ -1861,12 +1886,11 @@ func (n *ValueNode) SetValue(manager NodeManager, thisRef *NodeReference, this s
 			newNode := newHandle.Get().(*ValueNode)
 			newNode.key = n.key
 			newNode.value = value
-			newNode.hashDirty = true
+			newNode.markDirty()
 			newNode.pathLength = n.pathLength
 			return newRef, false, nil
 		}
 		n.value = value
-		n.hashDirty = true
 		n.markDirty()
 		return *thisRef, true, nil
 	}
@@ -1885,7 +1909,7 @@ func (n *ValueNode) SetValue(manager NodeManager, thisRef *NodeReference, this s
 	sibling := siblingHandle.Get().(*ValueNode)
 	sibling.key = key
 	sibling.value = value
-	sibling.hashDirty = true
+	sibling.markDirty()
 
 	thisPath := KeyToNibblePath(n.key, manager)
 	newRootId, err := splitLeafNode(manager, thisRef, thisPath[:], n, this, path, &siblingRef, sibling, siblingHandle)
@@ -1921,12 +1945,11 @@ func (n *ValueNode) setPathLength(manager NodeManager, thisRef *NodeReference, t
 		newNode := newHandle.Get().(*ValueNode)
 		newNode.key = n.key
 		newNode.value = n.value
-		newNode.hashDirty = true
+		newNode.markDirty()
 		newNode.pathLength = length
 		return newRef, false, nil
 	}
 
-	n.hashDirty = true
 	n.pathLength = length
 	n.markDirty()
 	return *thisRef, true, nil
@@ -1942,10 +1965,10 @@ func (n *ValueNode) Check(source NodeSource, thisRef *NodeReference, path []Nibb
 	//  - value must not be empty
 	//  - values are in the right position of the trie
 	//  - the path length is correct (if enabled to be tracked)
-	errs := []error{}
+	var errs []error
 
-	if !n.IsDirty() && n.hashDirty {
-		errs = append(errs, fmt.Errorf("node %v is marked clean but hash is dirty", thisRef.Id()))
+	if err := n.nodeBase.check(thisRef); err != nil {
+		errs = append(errs, err)
 	}
 
 	fullPath := KeyToNibblePath(n.key, source)
@@ -1967,7 +1990,7 @@ func (n *ValueNode) Check(source NodeSource, thisRef *NodeReference, path []Nibb
 }
 
 func (n *ValueNode) Dump(out io.Writer, source NodeSource, thisRef *NodeReference, indent string) error {
-	fmt.Fprintf(out, "%sValue (ID: %v/%t/%d, Hash: %v, dirtyHash: %t): %v - %v\n", indent, thisRef.Id(), n.IsFrozen(), n.pathLength, formatHashForDump(n.hash), n.hashDirty, n.key, n.value)
+	fmt.Fprintf(out, "%sValue (ID: %v/%t/%d, Hash: %v, hashState: %v): %v - %v\n", indent, thisRef.Id(), n.IsFrozen(), n.pathLength, formatHashForDump(n.hash), n.getHashStatus(), n.key, n.value)
 	return nil
 }
 
@@ -1993,7 +2016,7 @@ func (BranchNodeEncoderWithNodeHash) GetEncodedSize() int {
 }
 
 func (BranchNodeEncoderWithNodeHash) Store(dst []byte, node *BranchNode) error {
-	if node.hashDirty {
+	if !node.hasCleanHash() {
 		panic("unable to store branch node with dirty hash")
 	}
 	encoder := NodeIdEncoder{}
@@ -2016,6 +2039,7 @@ func (BranchNodeEncoderWithNodeHash) Load(src []byte, node *BranchNode) error {
 	}
 	src = src[step*16:]
 	copy(node.hash[:], src)
+	node.hashStatus = hashStatusClean
 
 	// The hashes of the child nodes are not stored with the node, so they are
 	// marked as dirty to trigger a re-computation the next time they are used.
@@ -2068,8 +2092,8 @@ func (BranchNodeEncoderWithChildHashes) Load(src []byte, node *BranchNode) error
 	}
 	node.embeddedChildren = binary.BigEndian.Uint16(src)
 
-	// The node's hash is not stored with the node, so it is marked dirty.
-	node.hashDirty = true
+	// The node's hash is not stored with the node, so it is marked unknown.
+	node.hashStatus = hashStatusUnknown
 
 	return nil
 }
@@ -2083,7 +2107,7 @@ func (ExtensionNodeEncoderWithNodeHash) GetEncodedSize() int {
 }
 
 func (ExtensionNodeEncoderWithNodeHash) Store(dst []byte, value *ExtensionNode) error {
-	if value.hashDirty {
+	if !value.hasCleanHash() {
 		panic("unable to store extension node with dirty hash")
 	}
 	pathEncoder := PathEncoder{}
@@ -2106,6 +2130,7 @@ func (ExtensionNodeEncoderWithNodeHash) Load(src []byte, node *ExtensionNode) er
 	node.next = NewNodeReference(id)
 	src = src[idEncoder.GetEncodedSize():]
 	copy(node.hash[:], src)
+	node.hashStatus = hashStatusClean
 
 	// The hash of the next node is not stored with the node, so it is marked
 	// as dirty to trigger a re-computation the next time it is accessed.
@@ -2155,8 +2180,8 @@ func (ExtensionNodeEncoderWithChildHash) Load(src []byte, node *ExtensionNode) e
 	src = src[common.HashSize:]
 	node.nextIsEmbedded = src[0] != 0
 
-	// The node's hash is not stored with the node, so it is marked dirty.
-	node.hashDirty = true
+	// The node's hash is not stored with the node, so it is marked unknown.
+	node.hashStatus = hashStatusUnknown
 
 	return nil
 }
@@ -2171,7 +2196,7 @@ func (AccountNodeEncoderWithNodeHash) GetEncodedSize() int {
 }
 
 func (AccountNodeEncoderWithNodeHash) Store(dst []byte, node *AccountNode) error {
-	if node.hashDirty {
+	if !node.hasCleanHash() {
 		panic("unable to store account node with dirty hash")
 	}
 	copy(dst, node.address[:])
@@ -2202,6 +2227,7 @@ func (AccountNodeEncoderWithNodeHash) Load(src []byte, node *AccountNode) error 
 	node.storage = NewNodeReference(id)
 	src = src[idEncoder.GetEncodedSize():]
 	copy(node.hash[:], src)
+	node.hashStatus = hashStatusClean
 
 	// The storage hash is not stored with the node, so it is marked as dirty to
 	// trigger a re-computation the next time it is accessed.
@@ -2252,8 +2278,8 @@ func (AccountNodeEncoderWithChildHash) Load(src []byte, node *AccountNode) error
 	src = src[idEncoder.GetEncodedSize():]
 	copy(node.storageHash[:], src)
 
-	// The node's hash is not stored with the node, so it is marked dirty.
-	node.hashDirty = true
+	// The node's hash is not stored with the node, so it is marked unknown.
+	node.hashStatus = hashStatusUnknown
 
 	return nil
 }
@@ -2312,8 +2338,8 @@ func (ValueNodeEncoderWithoutNodeHash) Load(src []byte, node *ValueNode) error {
 	src = src[len(node.key):]
 	copy(node.value[:], src)
 
-	// The node's hash is not stored with the node, so it is marked dirty.
-	node.hashDirty = true
+	// The node's hash is not stored with the node, so it is marked unknown.
+	node.hashStatus = hashStatusUnknown
 
 	return nil
 }
@@ -2325,6 +2351,9 @@ func (ValueNodeEncoderWithNodeHash) GetEncodedSize() int {
 }
 
 func (ValueNodeEncoderWithNodeHash) Store(dst []byte, node *ValueNode) error {
+	if !node.hasCleanHash() {
+		panic("unable to store value node with dirty hash")
+	}
 	ValueNodeEncoderWithoutNodeHash{}.Store(dst, node)
 	dst = dst[ValueNodeEncoderWithoutNodeHash{}.GetEncodedSize():]
 	copy(dst, node.hash[:])
@@ -2335,7 +2364,7 @@ func (ValueNodeEncoderWithNodeHash) Load(src []byte, node *ValueNode) error {
 	ValueNodeEncoderWithoutNodeHash{}.Load(src, node)
 	src = src[ValueNodeEncoderWithoutNodeHash{}.GetEncodedSize():]
 	copy(node.hash[:], src)
-	node.hashDirty = false
+	node.hashStatus = hashStatusClean
 	return nil
 }
 

--- a/go/state/mpt/nodes_test.go
+++ b/go/state/mpt/nodes_test.go
@@ -1154,6 +1154,7 @@ func TestBranchNode_VisitAbortByChild(t *testing.T) {
 }
 
 func TestBranchNode_CheckDetectsIssues(t *testing.T) {
+	unknownHashStatus := hashStatusUnknown
 	tests := map[string]struct {
 		setup NodeDesc
 		ok    bool
@@ -1206,6 +1207,11 @@ func TestBranchNode_CheckDetectsIssues(t *testing.T) {
 			dirty:            true,
 			dirtyHash:        false,
 			dirtyChildHashes: []int{1},
+		}, false},
+		"dirts_node_with_unknown_hash_status": {&Branch{
+			children:   Children{1: &Value{}, 2: &Value{}},
+			dirty:      true,
+			hashStatus: &unknownHashStatus,
 		}, false},
 	}
 	for name, test := range tests {
@@ -2396,20 +2402,22 @@ func TestExtensionNode_VisitAbortByChild(t *testing.T) {
 }
 
 func TestExtensionNode_CheckDetectsIssues(t *testing.T) {
+	unknownHashStatus := hashStatusUnknown
 	tests := map[string]struct {
 		setup NodeDesc
 		ok    bool
 	}{
-		"ok":                              {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}}, true},
-		"empty path":                      {&Extension{next: &Branch{}}, false},
-		"next not a branch":               {&Extension{path: []Nibble{1, 2, 3}, next: &Value{}}, false},
-		"invalid hash":                    {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, nextHash: common.Hash{1}}, false},
-		"dirty hash is ignored":           {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, nextHash: common.Hash{1}, dirty: true, hashDirty: true, nextHashDirty: true}, true},
-		"full_frozen":                     {&Extension{frozen: true, path: []Nibble{1, 2, 3}, next: &Branch{frozen: true}}, true},
-		"partial_frozen":                  {&Extension{frozen: false, path: []Nibble{1, 2, 3}, next: &Branch{frozen: true}}, true},
-		"inconsistent_frozen":             {&Extension{frozen: true, path: []Nibble{1, 2, 3}, next: &Branch{frozen: false}}, false},
-		"clean_with_dirty_hash":           {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, dirty: false, hashDirty: true}, false},
-		"clean_hash_with_dirty_next_hash": {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, dirty: false, hashDirty: false, nextHashDirty: true}, false},
+		"ok":                                  {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}}, true},
+		"empty path":                          {&Extension{next: &Branch{}}, false},
+		"next not a branch":                   {&Extension{path: []Nibble{1, 2, 3}, next: &Value{}}, false},
+		"invalid hash":                        {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, nextHash: common.Hash{1}}, false},
+		"dirty hash is ignored":               {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, nextHash: common.Hash{1}, dirty: true, hashDirty: true, nextHashDirty: true}, true},
+		"full_frozen":                         {&Extension{frozen: true, path: []Nibble{1, 2, 3}, next: &Branch{frozen: true}}, true},
+		"partial_frozen":                      {&Extension{frozen: false, path: []Nibble{1, 2, 3}, next: &Branch{frozen: true}}, true},
+		"inconsistent_frozen":                 {&Extension{frozen: true, path: []Nibble{1, 2, 3}, next: &Branch{frozen: false}}, false},
+		"clean_with_dirty_hash":               {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, dirty: false, hashDirty: true}, false},
+		"clean_hash_with_dirty_next_hash":     {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, dirty: false, hashDirty: false, nextHashDirty: true}, false},
+		"dirts_node_with_unknown_hash_status": {&Extension{path: []Nibble{1, 2, 3}, next: &Branch{}, dirty: true, hashStatus: &unknownHashStatus}, false},
 	}
 	for name, test := range tests {
 		t.Run(name, func(t *testing.T) {
@@ -3592,6 +3600,7 @@ func TestAccountNode_VisitAbortByChild(t *testing.T) {
 }
 
 func TestAccountNode_CheckDetectsIssues(t *testing.T) {
+	unknownHashStatus := hashStatusUnknown
 	tests := map[string]struct {
 		path  []Nibble
 		setup NodeDesc
@@ -3654,6 +3663,13 @@ func TestAccountNode_CheckDetectsIssues(t *testing.T) {
 			storage:          &Value{frozen: false},
 			dirtyHash:        false,
 			storageHashDirty: true,
+		}, false},
+		"dirts_node_with_unknown_hash_status": {[]Nibble{1, 2, 3}, &Account{
+			address:    common.Address{0x12, 0x34},
+			info:       AccountInfo{Nonce: common.Nonce{1}},
+			pathLength: 37,
+			dirty:      true,
+			hashStatus: &unknownHashStatus,
 		}, false},
 	}
 	for name, test := range tests {
@@ -4361,6 +4377,7 @@ func TestValueNode_Visit(t *testing.T) {
 }
 
 func TestValueNode_CheckDetectsIssues(t *testing.T) {
+	unknownHashStatus := hashStatusUnknown
 	tests := map[string]struct {
 		path  []Nibble
 		setup NodeDesc
@@ -4392,6 +4409,13 @@ func TestValueNode_CheckDetectsIssues(t *testing.T) {
 			length:    37,
 			dirty:     false,
 			dirtyHash: true,
+		}, false},
+		"dirts_node_with_unknown_hash_status": {[]Nibble{1, 2, 3}, &Value{
+			key:        common.Key{0x12, 0x34},
+			value:      common.Value{1},
+			length:     61,
+			dirty:      true,
+			hashStatus: &unknownHashStatus,
 		}, false},
 	}
 	for name, test := range tests {
@@ -4510,13 +4534,36 @@ func TestCheckForest_DetectsInvalidReUse(t *testing.T) {
 }
 
 // ----------------------------------------------------------------------------
+//                              HashStatus
+// ----------------------------------------------------------------------------
+
+func TestHashStatus_Print(t *testing.T) {
+	tests := []struct {
+		status hashStatus
+		print  string
+	}{
+		{hashStatusClean, "clean"},
+		{hashStatusDirty, "dirty"},
+		{hashStatusUnknown, "unknown"},
+		{hashStatus(12), "unknown"},
+	}
+
+	for _, test := range tests {
+		if want, got := test.print, test.status.String(); want != got {
+			t.Errorf("unexpected print, wanted %v, got %v", want, got)
+		}
+	}
+}
+
+// ----------------------------------------------------------------------------
 //                               Encoders
 // ----------------------------------------------------------------------------
 
 func TestAccountNodeEncoderWithNodeHash(t *testing.T) {
 	node := AccountNode{
 		nodeBase: nodeBase{
-			hash: common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+			hash:       common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+			hashStatus: hashStatusClean,
 		},
 		info: AccountInfo{
 			Nonce:    common.Nonce{1, 2, 3, 4, 5, 6, 7, 8},
@@ -4551,7 +4598,7 @@ func TestAccountNodeEncoderWithChildHash(t *testing.T) {
 	encoder.Store(buffer, &node)
 	recovered := AccountNode{}
 	encoder.Load(buffer, &recovered)
-	node.hashDirty = true
+	node.hashStatus = hashStatusUnknown
 	if !reflect.DeepEqual(node, recovered) {
 		t.Errorf("encoding/decoding failed, wanted %v, got %v", node, recovered)
 	}
@@ -4559,7 +4606,8 @@ func TestAccountNodeEncoderWithChildHash(t *testing.T) {
 func TestAccountNodeWithPathLengthEncoderWithNodeHash(t *testing.T) {
 	node := AccountNode{
 		nodeBase: nodeBase{
-			hash: common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+			hash:       common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+			hashStatus: hashStatusClean,
 		},
 		info: AccountInfo{
 			Nonce:    common.Nonce{1, 2, 3, 4, 5, 6, 7, 8},
@@ -4608,7 +4656,7 @@ func TestBranchNodeEncoderWithChildHashes(t *testing.T) {
 	encoder.Store(buffer, &node)
 	recovered := BranchNode{}
 	encoder.Load(buffer, &recovered)
-	node.hashDirty = true
+	node.hashStatus = hashStatusUnknown
 	if !reflect.DeepEqual(node, recovered) {
 		t.Errorf("encoding/decoding failed, wanted %v, got %v", node, recovered)
 	}
@@ -4617,7 +4665,8 @@ func TestBranchNodeEncoderWithChildHashes(t *testing.T) {
 func TestBranchNodeEncoderWithNodeHash(t *testing.T) {
 	node := BranchNode{
 		nodeBase: nodeBase{
-			hash: common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+			hash:       common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16},
+			hashStatus: hashStatusClean,
 		},
 		children: [16]NodeReference{
 			NewNodeReference(1),
@@ -4666,7 +4715,7 @@ func TestExtensionNodeEncoderWithChildHash(t *testing.T) {
 	encoder.Store(buffer, &node)
 	recovered := ExtensionNode{}
 	encoder.Load(buffer, &recovered)
-	node.hashDirty = true
+	node.hashStatus = hashStatusUnknown
 	if !reflect.DeepEqual(node, recovered) {
 		t.Errorf("encoding/decoding failed, wanted %v, got %v", node, recovered)
 	}
@@ -4674,7 +4723,8 @@ func TestExtensionNodeEncoderWithChildHash(t *testing.T) {
 func TestExtensionNodeEncoderWithNodeHash(t *testing.T) {
 	node := ExtensionNode{
 		nodeBase: nodeBase{
-			hash: common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+			hash:       common.Hash{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
+			hashStatus: hashStatusClean,
 		},
 		path: Path{
 			path:   [32]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
@@ -4705,7 +4755,7 @@ func TestValueNodeEncoderWithoutNodeHash(t *testing.T) {
 	encoder.Store(buffer, &node)
 	recovered := ValueNode{}
 	encoder.Load(buffer, &recovered)
-	node.hashDirty = true
+	node.hashStatus = hashStatusUnknown
 	if !reflect.DeepEqual(node, recovered) {
 		t.Errorf("encoding/decoding failed, wanted %v, got %v", node, recovered)
 	}
@@ -4714,7 +4764,8 @@ func TestValueNodeEncoderWithoutNodeHash(t *testing.T) {
 func TestValueNodeEncoderWithNodeHash(t *testing.T) {
 	node := ValueNode{
 		nodeBase: nodeBase{
-			hash: common.Hash{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31},
+			hash:       common.Hash{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31},
+			hashStatus: hashStatusClean,
 		},
 		key:   common.Key{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
 		value: common.Value{2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33},
@@ -4740,7 +4791,7 @@ func TestValueNodeWithPathLengthEncoderWithoutNodeHash(t *testing.T) {
 	encoder.Store(buffer, &node)
 	recovered := ValueNode{}
 	encoder.Load(buffer, &recovered)
-	node.hashDirty = true
+	node.hashStatus = hashStatusUnknown
 	if !reflect.DeepEqual(node, recovered) {
 		t.Errorf("encoding/decoding failed, wanted %v, got %v", node, recovered)
 	}
@@ -4749,7 +4800,8 @@ func TestValueNodeWithPathLengthEncoderWithoutNodeHash(t *testing.T) {
 func TestValueNodeWithPathLengthEncoderWithNodeHash(t *testing.T) {
 	node := ValueNode{
 		nodeBase: nodeBase{
-			hash: common.Hash{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31},
+			hash:       common.Hash{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31},
+			hashStatus: hashStatusClean,
 		},
 		key:        common.Key{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32},
 		value:      common.Value{2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33},
@@ -6426,16 +6478,12 @@ func markModifiedAsDirty(t *testing.T, ctxt *nodeContext, before, after NodeRefe
 			switch n := n.(type) {
 			case (*AccountNode):
 				n.markDirty()
-				n.hashDirty = true
 			case (*BranchNode):
 				n.markDirty()
-				n.hashDirty = true
 			case (*ExtensionNode):
 				n.markDirty()
-				n.hashDirty = true
 			case (*ValueNode):
 				n.markDirty()
-				n.hashDirty = true
 			}
 		}
 		// Also update the dirty child-hash markers in the nodes.
@@ -7361,6 +7409,7 @@ type Account struct {
 	storage          NodeDesc
 	storageHashDirty bool
 	dirtyHash        bool
+	hashStatus       *hashStatus // overrides dirtyHash flag if set
 }
 
 func (a *Account) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) {
@@ -7369,11 +7418,18 @@ func (a *Account) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) 
 		id, _ := ctx.Build(a.storage)
 		storage = id
 	}
+	hashStatus := hashStatusClean
+	if a.dirtyHash {
+		hashStatus = hashStatusDirty
+	}
+	if a.hashStatus != nil {
+		hashStatus = *a.hashStatus
+	}
 	return NewNodeReference(AccountId(ctx.nextIndex())), shared.MakeShared[Node](&AccountNode{
 		nodeBase: nodeBase{
-			clean:     !a.dirty,
-			frozen:    a.frozen,
-			hashDirty: a.dirtyHash,
+			clean:      !a.dirty,
+			frozen:     a.frozen,
+			hashStatus: hashStatus,
 		},
 		address:          a.address,
 		info:             a.info,
@@ -7394,6 +7450,7 @@ type Branch struct {
 	frozen           bool
 	frozenChildren   []int
 	dirtyHash        bool
+	hashStatus       *hashStatus // overrides dirtyHash flag if set
 }
 
 func (b *Branch) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) {
@@ -7414,7 +7471,13 @@ func (b *Branch) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) {
 	for _, i := range b.frozenChildren {
 		res.setChildFrozen(byte(i), true)
 	}
-	res.hashDirty = b.dirtyHash
+	res.hashStatus = hashStatusClean
+	if b.dirtyHash {
+		res.hashStatus = hashStatusDirty
+	}
+	if b.hashStatus != nil {
+		res.hashStatus = *b.hashStatus
+	}
 	return ref, shared.MakeShared[Node](res)
 }
 
@@ -7426,6 +7489,7 @@ type Extension struct {
 	hashDirty     bool
 	nextHash      common.Hash
 	nextHashDirty bool
+	hashStatus    *hashStatus // overrides dirtyHash flag if set
 }
 
 func (e *Extension) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) {
@@ -7435,7 +7499,13 @@ func (e *Extension) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]
 	res.frozen = e.frozen
 	res.path = CreatePathFromNibbles(e.path)
 	res.next, _ = ctx.Build(e.next)
-	res.hashDirty = e.hashDirty
+	res.hashStatus = hashStatusClean
+	if e.hashDirty {
+		res.hashStatus = hashStatusDirty
+	}
+	if e.hashStatus != nil {
+		res.hashStatus = *e.hashStatus
+	}
 	res.nextHash = e.nextHash
 	res.nextHashDirty = e.nextHashDirty
 	return ref, shared.MakeShared[Node](res)
@@ -7453,20 +7523,28 @@ func (t *Tag) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) {
 }
 
 type Value struct {
-	dirty     bool
-	key       common.Key
-	value     common.Value
-	length    byte
-	dirtyHash bool
-	frozen    bool
+	dirty      bool
+	key        common.Key
+	value      common.Value
+	length     byte
+	dirtyHash  bool
+	frozen     bool
+	hashStatus *hashStatus // overrides dirtyHash flag if set
 }
 
 func (v *Value) Build(ctx *nodeContext) (NodeReference, *shared.Shared[Node]) {
+	hashStatus := hashStatusClean
+	if v.dirtyHash {
+		hashStatus = hashStatusDirty
+	}
+	if v.hashStatus != nil {
+		hashStatus = *v.hashStatus
+	}
 	return NewNodeReference(ValueId(ctx.nextIndex())), shared.MakeShared[Node](&ValueNode{
 		nodeBase: nodeBase{
-			clean:     !v.dirty,
-			frozen:    v.frozen,
-			hashDirty: v.dirtyHash,
+			clean:      !v.dirty,
+			frozen:     v.frozen,
+			hashStatus: hashStatus,
 		},
 		key:        v.key,
 		value:      v.value,
@@ -7813,8 +7891,8 @@ func (c *nodeContext) diff(prefix string, nodeA, nodeB Node) []string {
 			if a.info != b.info {
 				diffs = append(diffs, fmt.Sprintf("%s: different info, got %v and %v", prefix, a.info, b.info))
 			}
-			if a.hashDirty != b.hashDirty {
-				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %t and %t", prefix, a.hashDirty, b.hashDirty))
+			if a.hashStatus != b.hashStatus {
+				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %v and %v", prefix, a.hashStatus, b.hashStatus))
 			}
 			if a.storageHashDirty != b.storageHashDirty {
 				diffs = append(diffs, fmt.Sprintf("%s: different storage hash-dirty flag, got %t and %t", prefix, a.storageHashDirty, b.storageHashDirty))
@@ -7840,8 +7918,8 @@ func (c *nodeContext) diff(prefix string, nodeA, nodeB Node) []string {
 			if a.path != b.path {
 				diffs = append(diffs, fmt.Sprintf("%s: different extension path, got %v and %v", prefix, a.path, b.path))
 			}
-			if a.hashDirty != b.hashDirty {
-				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %t and %t", prefix, a.hashDirty, b.hashDirty))
+			if a.hashStatus != b.hashStatus {
+				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %v and %v", prefix, a.hashStatus, b.hashStatus))
 			}
 			if a.nextHashDirty != b.nextHashDirty {
 				diffs = append(diffs, fmt.Sprintf("%s: different next-hash-dirty flag, got %t and %t", prefix, a.nextHashDirty, b.nextHashDirty))
@@ -7860,8 +7938,8 @@ func (c *nodeContext) diff(prefix string, nodeA, nodeB Node) []string {
 			if a.frozenChildren != b.frozenChildren {
 				diffs = append(diffs, fmt.Sprintf("%s: different frozen children flags, got %016b and %016b", prefix, a.frozenChildren, b.frozenChildren))
 			}
-			if a.hashDirty != b.hashDirty {
-				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %t and %t", prefix, a.hashDirty, b.hashDirty))
+			if a.hashStatus != b.hashStatus {
+				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %v and %v", prefix, a.hashStatus, b.hashStatus))
 			}
 			if a.dirtyHashes != b.dirtyHashes {
 				diffs = append(diffs, fmt.Sprintf("%s: different dirty-child-hashes flags, got %016b and %016b", prefix, a.dirtyHashes, b.dirtyHashes))
@@ -7880,8 +7958,8 @@ func (c *nodeContext) diff(prefix string, nodeA, nodeB Node) []string {
 			if a.frozen != b.frozen {
 				diffs = append(diffs, fmt.Sprintf("%s: different frozen state, got %t and %t", prefix, a.frozen, b.frozen))
 			}
-			if a.hashDirty != b.hashDirty {
-				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %t and %t", prefix, a.hashDirty, b.hashDirty))
+			if a.hashStatus != b.hashStatus {
+				diffs = append(diffs, fmt.Sprintf("%s: different hash-dirty flag, got %v and %v", prefix, a.hashStatus, b.hashStatus))
 			}
 			if a.key != b.key {
 				diffs = append(diffs, fmt.Sprintf("%s: different key, got %x and %x", prefix, a.key, b.key))
@@ -7936,7 +8014,7 @@ func (c *nodeContext) equalWithConfig(a, b Node, config equalityConfig) bool {
 		if b, ok := b.(*AccountNode); ok {
 			eq := a.address == b.address
 			eq = eq && a.info == b.info
-			eq = eq && (config.ignoreDirtyHash || a.hashDirty == b.hashDirty)
+			eq = eq && (config.ignoreDirtyHash || a.hashStatus == b.hashStatus)
 			eq = eq && (config.ignoreDirtyHash || a.storageHashDirty == b.storageHashDirty)
 			eq = eq && (config.ignoreFreeze || a.frozen == b.frozen)
 			eq = eq && c.equalTriesWithConfig(a.storage, b.storage, config)
@@ -7956,7 +8034,7 @@ func (c *nodeContext) equalWithConfig(a, b Node, config equalityConfig) bool {
 	if a, ok := a.(*ExtensionNode); ok {
 		if b, ok := b.(*ExtensionNode); ok {
 			eq := a.path == b.path
-			eq = eq && (config.ignoreDirtyHash || a.hashDirty == b.hashDirty)
+			eq = eq && (config.ignoreDirtyHash || a.hashStatus == b.hashStatus)
 			eq = eq && (config.ignoreDirtyHash || a.nextHashDirty == b.nextHashDirty)
 			eq = eq && (config.ignoreFreeze || a.frozen == b.frozen)
 			eq = eq && c.equalTriesWithConfig(a.next, b.next, config)
@@ -7970,7 +8048,7 @@ func (c *nodeContext) equalWithConfig(a, b Node, config equalityConfig) bool {
 			if !config.ignoreFreeze && a.frozen != b.frozen {
 				return false
 			}
-			if a.hashDirty != b.hashDirty {
+			if !config.ignoreDirtyHash && a.hashStatus != b.hashStatus {
 				return false
 			}
 			if !config.ignoreDirtyHash && a.dirtyHashes != b.dirtyHashes {
@@ -7993,7 +8071,7 @@ func (c *nodeContext) equalWithConfig(a, b Node, config equalityConfig) bool {
 		if b, ok := b.(*ValueNode); ok {
 			eq := a.key == b.key
 			eq = eq && a.value == b.value
-			eq = eq && (config.ignoreDirtyHash || a.hashDirty == b.hashDirty)
+			eq = eq && (config.ignoreDirtyHash || a.hashStatus == b.hashStatus)
 			eq = eq && (config.ignoreFreeze || a.frozen == b.frozen)
 			if !eq {
 				return false

